src/mlpack/methods/ann/loss_functions/dice_loss_impl.hpp

```diff
@@ -0,0 +1,63 @@
+/**
+ * @file dice_loss_impl.hpp
+ * @author N Rajiv Vaidyanathan
+ *
+ * Implementation of the dice loss function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTIONS_DICE_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTIONS_DICE_LOSS_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "dice_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+DiceLoss<InputDataType, OutputDataType>::DiceLoss(
+    const double smooth) : smooth(smooth)
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+double DiceLoss<InputDataType, OutputDataType>::Forward(
+    const InputType&& input, const TargetType&& target)
+{
+  return 1 - ((2 * arma::accu(target % input) + smooth) /
+    (arma::accu(target % target) + arma::accu(
+    input % input) + smooth));
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType, typename OutputType>
+void DiceLoss<InputDataType, OutputDataType>::Backward(
+    const InputType&& input,
+    const TargetType&& target,
+    OutputType&& output)
+{
+  output = -2 * (target * (arma::accu(input % input) +
+    arma::accu(target % target)) - 2 * input *
+    (arma::accu(target % input))) / std::pow(arma::accu(target % target)
+    + arma::accu(input % input), 2.0);
+}
```

Comment:
I think it also might be useful to include the smoothness parameter in the derivative.

Reply:
@zoq I'm not sure I understand. What does 'check the derivative' mean? My understanding is that if you are using a smooth factor to prevent the coefficient from getting to zero, and you don't use the same for its derivative, then there's no guarantee that the derivative wouldn't vanish or explode. Hence the suggestion, also it seems plausible to use in this situation, since the denominator is practically the same as the coefficient, just raised to the power of two.

Reply:
Yeah it makes sense to add the `smooth` parameter to prevent the gradient from blowing up... Also the main purpose of the parameter is to prevent overfitting so it might help in the gradient as well. You check it out [here](https://github.com/pytorch/pytorch/issues/1249#issuecomment-337999895).

@zoq @ShikharJ  I have updated the gradient in the latest commit. Please let me know if it is fine. I checked  the tests and it seemed to work fine.

Reply:
@zoq Yeah I checked, it's approximately the same.

Reply:
The only issue is that the gradient isn't `0` when the `input=target` as the `smooth` keeps the gradient `non-zero`.

Reply:
@Rajiv2605 In that case, would you mind adding an `eps` parameter (and include it only for the denominators) for the coefficient and the derivative? This is done in other classes as well, you may set the default value from there. That should take care of the issue.

Reply:
The derivative could be wrong, wolframalpha might be helpful here; example https://www.wolframalpha.com/input/?i=d%2Fdb+a(b%2Bx)%5E(-1%2Fc)

Reply:
@zoq The derivative is right... I got the same value in wolframalpha

Reply:
Do you mind to share the link to the wolframalpha link?

Reply:
> Do you mind to share the link to the wolframalpha link?

Link: https://www.wolframalpha.com/input/?i=d%2Fdb+2ab%2F(b%5E2+%2B+a%5E2)

Reply:
@ShikharJ The tests were made just by referring to the other tests in the file.

Reply:
@zoq @ShikharJ Why do you feel the derivative might be wrong?

Reply:
@Rajiv2605 thanks for the link, will take a look at the PR tomorrow, but if the derivative is right, not saying it's not, I think there is nothing else to do here. Don't think we have to check some other implementation to be sure the method is correct.

Reply:
@Rajiv2605 I meant a source for the numerical tests. What made you think that the expected value of the result is the one that is mentioned in the comments?

Reply:
I tried calculating manually and also checked by printing the test outputs... Both were approximately the same, maybe differing in the second or the third decimal place... That's how I got those values.

Reply:
So I added the smooth factor here: https://www.wolframalpha.com/input/?i=d%2Fdb+(2ab+%2B+s)+%2F+(b%5E2+%2B+a%5E2+%2B+s), and the denominator is the same, but the numerator is different, @Rajiv2605 do you see any issue on my side e.g. regarding the input function?

Reply:
@zoq It's right... The numerator looks different as it doesn't have the summation symbol.

Reply:
@ShikharJ @zoq I checked in Pytorch and I get similar values for gradients. Link: https://colab.research.google.com/drive/15XJBFRV8tzyD147nVVMwqm_h9QvuiEIz

Reply:
But if I set ``N = 1`` there is no summation.

Reply:
Just checked the notebook, ``-0.0878`` is somewhat different from the value (``-0.08230452``) we compare against.

Reply:
@zoq @ShikharJ Please check the latest commit... The `Backward` function is right now :+1: I'm getting the output I got in Pytorch.

Reply:
Personally, I don't use the results from another implementation as the main check, there is always the possibility the implementation isn't correct or they do something different as we do; so I think checking the derivative is more insightful. Anyway, I rechecked the updated backward step and now the numerator matches as well: https://www.wolframalpha.com/input/?i=(a+*+(b%5E2+%2B+a%5E2+%2B+s)+-+b+*+(2+*+a+*+b+%2B+s))

Reply:
> Personally, I don't use the results from another implementation as the main check, there is always the possibility the implementation isn't correct or they do something different as we do; so I think checking the derivative is more insightful. Anyway, I rechecked the updated backward step and now the numerator matches as well: https://www.wolframalpha.com/input/?i=(a+*+(b%5E2+%2B+a%5E2+%2B+s)+-+b+*+(2+*+a+*+b+%2B+s))

Yeah, in the numerator, I had to make a small correction... The `2` should be inside the paranthesis but it was outside before. Hence, it multiplied `smooth` by `2` as well and that was the mistake :)